herd/BPFSem.ml

```diff
@@ -0,0 +1,249 @@
+  module Act = MachAction.Make (C.PC) (BPF)
+  include SemExtra.Make (C) (BPF) (Act)
+
+  (* Barrier pretty print *)
+  let sync = None
+  let barriers = []
+  let isync = None
+
+  (*  TODO: let nat_sz = MachSize.Quad (* 64-bit Registers *) *)
+  let nat_sz = V.Cst.Scalar.machsize
+  let atomic_pair_allowed _ _ = true
+
+  (********************)
+  (* Semantics proper *)
+  (********************)
+
+  module Mixed (SZ : ByteSize.S) = struct
+    let ( >>= ) = M.( >>= )
+    let ( >>*= ) = M.( >>*= )
+    let ( >>| ) = M.( >>| )
+    let ( >>! ) = M.( >>! )
+    let ( >>:: ) = M.( >>:: )
+    let unimplemented op = Warn.user_error "BPF operation %s is not implemented (yet)" op
+
+    let tr_opamo op =
+      match op with
+      | BPF.AMOXCHG -> assert false
+      | BPF.ADD -> Op.Add
+      | BPF.AND -> Op.And
+      | BPF.OR -> Op.Or
+      | BPF.XOR -> Op.Xor
+      | BPF.AMOCMPXCHG -> assert false
+      | _ -> unimplemented "atomic op"
+    ;;
+
+    let tr_op = function
+      | BPF.ADD -> Op.Add
+      | BPF.SUB -> Op.Sub
+      | BPF.AND -> Op.And
+      | BPF.OR -> Op.Or
+      | BPF.XOR -> Op.Xor
+      | BPF.MUL -> Op.Mul
+      | BPF.DIV -> Op.Div
+      | BPF.REM -> Op.Rem
+      | BPF.LSL -> Op.ShiftLeft
+      | BPF.LSR -> Op.Lsr
+      | BPF.ASR -> unimplemented (BPF.pp_op BPF.ASR)
+      | BPF.AMOCMPXCHG -> unimplemented "non-atomic CMPXCHG"
+      | BPF.AMOXCHG -> unimplemented "non-atomic XCHG"
+    ;;
+
+    let tr_cond cond =
+      match cond with
+      | BPF.EQ -> Op.Eq
+      | BPF.NE -> Op.Ne
+      | BPF.LT -> Op.Lt
+      | BPF.GE -> Op.Ge
+    ;;
+
+    let mk_read sz ato loc v =
+      Act.Access (Dir.R, loc, v, ato, (), sz, Act.access_of_location_std loc)
+    ;;
+
+    let read_reg is_data r ii =
+      M.read_loc is_data (mk_read nat_sz BPF.N) (A.Location_reg (ii.A.proc, r)) ii
+    ;;
+
+    let read_reg_ord = read_reg false
+    let read_reg_data = read_reg true
+
+    let do_read_mem sz ato a ii =
+      M.read_loc false (mk_read sz ato) (A.Location_global a) ii
+    ;;
+
+    let read_mem sz a ii = do_read_mem sz BPF.N a ii
+    let read_mem_atomic sz a ii = do_read_mem sz BPF.X a ii
+    let read_mem_sc sz a ii = do_read_mem sz BPF.SC a ii
+
+    let write_reg r v ii =
+      M.mk_singleton_es
+        (Act.Access
+           (Dir.W, A.Location_reg (ii.A.proc, r), v, BPF.N, (), nat_sz, Access.REG))
+        ii
+    ;;
+
+    let write_mem_sc sz a v ii =
+      M.mk_singleton_es
+        (Act.Access (Dir.W, A.Location_global a, v, BPF.SC, (), sz, Access.VIR))
+        ii
+    ;;
+
+    let write_mem sz a v ii =
+      M.mk_singleton_es
+        (Act.Access (Dir.W, A.Location_global a, v, BPF.N, (), sz, Access.VIR))
+        ii
+    ;;
+
+    let write_mem_atomic sz a v resa ii =
+      let eq = [ M.VC.Assign (a, M.VC.Atom resa) ] in
+      M.mk_singleton_es_eq
+        (Act.Access (Dir.W, A.Location_global a, v, BPF.X, (), sz, Access.VIR))
+        eq
+        ii
+    ;;
```

Comment:
This function `write_mem_atomic` is not called.

Reply:
Fixed in next version